### ISSUE TITLE
Dart Project deploy to Firebase

### ISSUE THREAD

**Issue by Harikrishnadhas** [travis-ci/travis-ci#8131]
I'm trying to implement Travis CI for building my Angular Dart Project and deploying to Firebase.

It is going through the build and test fine, and while deploying i'm getting the below message.

> **Error: Specified public directory does not exist, can't deploy hosting**

**Git Hub:** https://github.com/Harikrishnadhas/siththargnaanaalayam
**Travis CI :** https://travis-ci.org/Harikrishnadhas/siththargnaanaalayam/builds/255561107

I have choosen build/web as my public directory while creating the firebase.json file. Seems like build directory is a different one in Travis CI. 

What is the build directory?
Can i see some build files done in Travis CI for my project?
If so how?

**Comment by BanzaiMan** [comment]
If `build/web` directory is not a part of the repository, that will be cleaned up before deployment. You can see the message here: https://travis-ci.org/Harikrishnadhas/siththargnaanaalayam/builds/255561107#L1316

> Cleaning up git repository with `git stash --all`. If you need build artifacts for deployment, set `deploy.skip_cleanup: true`. See https://docs.travis-ci.com/user/deployment/#Uploading-Files.

**Comment by Harikrishnadhas** [comment]
Done that. Now it is not cleaning up. But Still same issue persists. Please see below.

https://travis-ci.org/Harikrishnadhas/siththargnaanaalayam/jobs/255648371#L1324

**Comment by Harikrishnadhas** [comment]
[](url)https://travis-ci.org/Harikrishnadhas/siththargnaanaalayam/jobs/255648371#L1324

**Comment by Harikrishnadhas** [comment]
Will Travis CI call pub build? or we should configure that any where? because if i do pub build from my development machine, i get the below log.

_```
Working dir: /mnt/0A16E9EE16E9DAA9/Projects/siththargnaanaalayam
/mnt/0A16E9EE16E9DAA9/dart-sdk/bin/pub build --mode=release --output=build
Loading source assets...
Loading dart_to_js_script_rewriter and angular2/transform/codegen transformers...
Loading angular2 transformers...
Building WebUI...
[Info from Dart2JS]:
Compiling WebUI|web/main.dart...
[Info from Dart2JS]:
Took 0:00:17.170424 to compile WebUI|web/main.dart.
Built 1544 files to "build".
Process finished with exit code 0
```_

In Travis CI log, i don't see these lines.

**Comment by BanzaiMan** [comment]
Where in this build https://travis-ci.org/Harikrishnadhas/siththargnaanaalayam/jobs/255648371 `pub/web` is supposed to be generated? I don't know much about Dart tool chain or Firebase, but https://github.com/Harikrishnadhas/siththargnaanaalayam/blob/e1f4e665a9ad0a0d24e0cbbe9fc622d49f8ce260/.travis.yml doesn't seem to do very much.

**Comment by Harikrishnadhas** [comment]
It should be generating here https://travis-ci.org/Harikrishnadhas/siththargnaanaalayam/jobs/255648371#L595.

Travis should run the below command in the repository which will generate the build/web
`pub build`

**Comment by BanzaiMan** [comment]
Are you sure about that? I tried cloning your repository and running the build steps, but at the end, there is no `build` directory at the repository root.

Where in your configuration is `pub build` invoked? I don't think this is invoked by default, so you'll have to do that yourself. https://docs.travis-ci.com/user/languages/dart#Running-Tests

**Comment by Harikrishnadhas** [comment]
One basic doubt, what will be script executed by default for Dart?? I didn't given anything so far for script.

**Comment by BanzaiMan** [comment]
Beyond that, this is a Dart issue, so please mention community support members for advice.

**Comment by Harikrishnadhas** [comment]
Finally got it working. I just mentioned the below in the travis.yml.

`script: pub build --mode=release --output=build`

And now build process is happening, and the Travis uploads the built files to firebase successfully. See here https://travis-ci.org/Harikrishnadhas/siththargnaanaalayam/jobs/255667748#L600

But the half of the process seems to be repeating after adding the pub build command in script. Pub get also does the transformation and pub build also does that, which increases the build time.

Yes, this is to do with the Dart support for Travis, Hope community sees this, provides some straight forward solution, which doesn't duplicate any process.

**Comment by BanzaiMan** [comment]
Paging @nex3 @a14n

**Comment by Harikrishnadhas** [comment]
@BanzaiMan you should have downloaded the Dart SDK and run the pub build, then you will get the build folder.

**Comment by BanzaiMan** [comment]
@Harikrishnadhas I understand. But recall that your build was not configured to do that. That explains why your deployment was failing.

**Comment by Harikrishnadhas** [comment]
@BanzaiMan Yes my build in Travis was failing, but not the one in local. Also the below step, i just tried myself.

`script: pub build --mode=release --output=build`

It was not there in any doc for dart lang. Hence i assumed that Travis will take care of running pub build.

**Comment by Harikrishnadhas** [comment]
@nex3 @a14n can you comment on the below issue? The workaround which i have done to build the dart app is the right one or is there any other straight forward option available.

**script: pub build --mode=release --output=build**

**Comment by a14n** [comment]
The current Dart support for Travis does not do any `pub build`. [It handles several tasks](https://docs.travis-ci.com/user/languages/dart) but nothing specific to `deploy` phase.

I'm not used to the deploy travis support and I don't know the best pratices in that area but I think you should put your `script` in a `before_deploy` phase. Thus you will be able to use the `dart_task` for build and build the release only for deployment (BTW be careful how matrix expansion work with deploy).

**Comment by Harikrishnadhas** [comment]
@a14n Thanks for the clarification.
